### PR TITLE
Require the minimum JDK version to 11 for examples only

### PR DESCRIPTION
**Pull request by minwoox**
Motivation:
So that we can leverage JDK 9+ APIs such as `Map.of(...)`.

Modifications:
- Use 11 for javac `release` option.
- (Misc) Add `rootProject.ext.testJavaVersion` so that we can use it in any of the build.gradle files.

Result:
- You need to use at least JDK 11 for examples.

### CODE REVIEW

File: build.gradle [hunk 1]
```diff
@@ -157,9 +157,20 @@ configure(projectsWithFlags('java')) {
...
     }
 
-    // Target Java 8.
-    tasks.withType(JavaCompile) {
-        options.compilerArgs += ['--release', '8']
+    tasks.withType(JavaCompile) { task ->
+
+        if (task.path.startsWith(':examples:')) {
+            // Target Java 11 for examples.
+            options.compilerArgs += ['--release', '11']
```
Comment (anuraaga) [inline build.gradle:164]: `options.release = 11` or similar should work nowadays

Comment (minwoox) [reply build.gradle:164]: Thanks, let me try that. 😉

**Review by trustin — APPROVED** [review]
LGTM once @anuraaga's comment is addressed.

File: gradle/scripts/lib/java.gradle [hunk 2]
```diff
@@ -1,28 +1,36 @@
...
+    if (rootProject.hasProperty('testJavaVersion')) {
+        def testVersion = rootProject.findProperty('testJavaVersion')
+        if (buildJdkVersion != testVersion) {
+            testJavaVersion = testVersion
+            logger.quiet("Overriding JRE for tests with ${testJavaVersion}")
+        } else {
+            testJavaVersion = buildJdkVersion
+        }
+    } else {
+        testJavaVersion = buildJdkVersion
```
Comment (ikhoon) [inline gradle/scripts/lib/java.gradle:23]: Is it better to use int value than string value?

File: core/build.gradle [hunk 3]
```diff
@@ -41,9 +41,9 @@ mrJarVersions.each { version->
         testClassesDirs = sourceSets."java${version}Test".output.classesDirs
         classpath = sourceSets."java${version}Test".runtimeClasspath
 
-        enabled = (Integer.parseInt(JavaVersion.current().getMajorVersion()) >= version) &&
-                (!rootProject.hasProperty('testJavaVersion') ||
-                        Integer.parseInt(rootProject.findProperty('testJavaVersion')) >= version)
+        enabled = Integer.parseInt(JavaVersion.current().getMajorVersion()) >= version &&
+                rootProject.ext.testJavaVersion != null &&
```
Comment (ikhoon) [inline core/build.gradle:45]: IIUC, `rootProject.ext.testJavaVersion` is not null always?

**Review by ikhoon — APPROVED** [review]
Thanks! @minwoox. Left minor nits.

File: gradle/scripts/lib/java.gradle (hunk 2 above)
Comment (minwoox) [reply gradle/scripts/lib/java.gradle:23]: IIUC, we cannot define an integer global variable.
https://docs.gradle.org/current/dsl/org.gradle.api.plugins.ExtraPropertiesExtension.html

File: core/build.gradle (hunk 3 above)
Comment (minwoox) [reply core/build.gradle:45]: Yeah, it's always not null. 😄

File: gradle/scripts/lib/java.gradle (hunk 2 above)
Comment (ikhoon) [reply gradle/scripts/lib/java.gradle:23]: Thanks for the link. TIL. 😄

Comment (ikhoon) [reply gradle/scripts/lib/java.gradle:23]: By the way, Integer seems working. `ExtraPropertiesExtension.set(java.lang.String, java.lang.Object)`

Comment (minwoox) [reply gradle/scripts/lib/java.gradle]: Oh, it's working. Thanks! 😄

**Comment by minwoox** [comment]
Sent a PR to https://github.com/line/gradle-scripts/pull/100

**Review by trustin — APPROVED** [review]
Thanks, @minwoox!

**Pull request merged**